### PR TITLE
Drop Python3.8 support

### PR DESCRIPTION
**Pull request by stankudrow**
### Issue reference
No issue for dropping py3.8 support according to its [end of life](https://endoflife.date/python) date.

### Changes
Removing any dependencies on the Python v3.8, like in:
- the `pyproject.toml` file
- code
- CI/CD pipelines -> "Github Actions" workflows
- Makefile - just hit on occasion

### Checks and tests
`make all-checks`
make_all-checks

Partially relates to the PR #852 .

### CODE REVIEW

File: docs/conf.py [hunk 1]
```diff
@@ -10,9 +10,9 @@
 # All configuration values have a default; values that are commented out
 # serve to show the default.
 
-import sys, os
+import sys
+import os
 
-import sphinx_rtd_theme
```
Comment (stankudrow) [inline docs/conf.py:15]: From [pyupgrade](https://pypi.org/project/pyupgrade/) (seen from the PR #852).

**Comment by stankudrow** [comment]
@bbayles , hello.

I have some questions in advance and your helping hand is much appreciated:

- https://github.com/more-itertools/more-itertools/blob/master/.github/workflows/python-app.yml#L24-L35 - any guidelines for this block? I am new with GHA, time for immersion needed.
- https://github.com/more-itertools/more-itertools/blob/master/more_itertools/more.py#L243 - have not found the version according to this hex -> to retain or can be left out safely? Need time to figure out

File: Makefile [hunk 2]
```diff
@@ -1,35 +1,29 @@
...
-.PHONY: requirements
 requirements:
 	python3 -m pip install -r requirements/development.txt
 	python3 -m pip install --editable .
 
-.PHONY: check
 check:
 	black --check .
-	flake8 --exclude .venv,venv,docs/conf.py .
+	flake8 more_itertools tests
```
Comment (stankudrow) [inline Makefile:11]: Let's be more specific in targeting code for checks and assuming nothing out is of any interest.

File: more_itertools/more.py [hunk 3]
```diff
@@ -240,15 +240,14 @@ def last(iterable, default=_marker):
         if isinstance(iterable, Sequence):
             return iterable[-1]
         # Work around https://bugs.python.org/issue38525
-        elif hasattr(iterable, '__reversed__') and (hexversion != 0x030800F0):
```
Comment (stankudrow) [inline more_itertools/more.py:243]: No more need in checking the Py3.8 here.

**Comment by bbayles** [comment]
> https://github.com/more-itertools/more-itertools/blob/master/.github/workflows/python-app.yml#L24-L35 - any guidelines for this block? I am new with GHA, time for immersion needed.

I wanted to run these checks on the lowest supported version.

File: .github/workflows/python-app.yml [hunk 4]
```diff
@@ -22,17 +22,14 @@ jobs:
...
     - name: Build packages
-      if: "matrix.python-version == '3.8'"
       run: make package
     - name: Upload packages
-      if: "matrix.python-version == '3.8'"
       uses: actions/upload-artifact@v4
       with:
         name: more-itertools-packages
         path: dist/*
+        overwrite: true
```
Comment (stankudrow) [inline .github/workflows/python-app.yml:35]: Testing this workaround.

References:
- https://github.com/actions/upload-artifact/issues/506
- https://devops.stackexchange.com/questions/19966/what-happens-in-github-if-artifact-names-conflicts

**Comment by stankudrow** [comment]
> > https://github.com/more-itertools/more-itertools/blob/master/.github/workflows/python-app.yml#L24-L35 - any guidelines for this block? I am new with GHA, time for immersion needed.
> 
> I wanted to run these checks on the lowest supported version.

Ah, just replacing 3.8 with 3.9, is that right?

What about [this "workaround"](https://github.com/more-itertools/more-itertools/pull/928#discussion_r1888943770)? Could you share your suggestions?

File: .github/workflows/python-app.yml (hunk 4 above)
Comment (stankudrow) [reply .github/workflows/python-app.yml]: Removed, maybe temporarily, so not resolving now.

File: Makefile [hunk 5]
```diff
@@ -1,35 +1,29 @@
-.PHONY: all-checks
+.PHONY: all-checks check coverage docs format package requirements test
```
Comment (bbayles) [inline Makefile:1]: Please revert these changes; they're not related to the goal of deprecating 3.8.

File: more_itertools/more.py [hunk 6]
```diff
@@ -218,8 +218,8 @@ def first(iterable, default=_marker):
         return item
     if default is _marker:
         raise ValueError(
-            'first() was called on an empty iterable, and no '
-            'default value was provided.'
+            "first() was called on an empty iterable, "
```
Comment (bbayles) [inline more_itertools/more.py:221]: Please restore these quotes to their prior state.

File: Makefile (hunk 5 above)
Comment (stankudrow) [reply Makefile:1]: Are they OK for another PR or not at all?

File: more_itertools/more.py (hunk 6 above)
Comment (stankudrow) [reply more_itertools/more.py:221]: Will be done.

Comment (stankudrow) [reply more_itertools/more.py]: Single quotes are back as requested.

File: Makefile (hunk 5 above)
Comment (bbayles) [reply Makefile]: Is there some benefit? My sense is no - it's fine to do it either way.
